frontend/src/main/java/org/abs_models/frontend/typechecker/nullable/NullCheckerExtension.java

```diff
@@ -0,0 +1,287 @@
+package org.abs_models.frontend.typechecker.nullable;
+
+import org.abs_models.frontend.analyser.BitVec;
+import org.abs_models.frontend.analyser.ErrorMessage;
+import org.abs_models.frontend.analyser.SemanticWarning;
+import org.abs_models.frontend.analyser.TypeError;
+import org.abs_models.frontend.ast.*;
+import org.abs_models.frontend.typechecker.Type;
+import org.abs_models.frontend.typechecker.TypeAnnotation;
+import org.abs_models.frontend.typechecker.ext.AdaptDirection;
+import org.abs_models.frontend.typechecker.ext.DefaultTypeSystemExtension;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class NullCheckerExtension extends DefaultTypeSystemExtension {
+    public static String NULLABLE_KEY = "NULLABLE_KEY";
+    private NullableType defaultType = NullableType.Nullable;
+    private boolean warnAboutMissingAnnotation = false;
+    private boolean checkNullCall = false;
+
+    public NullCheckerExtension(Model m) {
+        super(m);
+    }
+
+    public NullCheckerExtension(Model m, NullableType defaultType) {
+        super(m);
+        this.defaultType = defaultType;
+    }
+
+    public void setDefaultType(NullableType nt) {
+        defaultType = nt;
+    }
+
+    public void setWarnAboutMissingAnnotation(boolean v) {
+        warnAboutMissingAnnotation = v;
+    }
+
+    public void checkNullCall() {
+        this.checkNullCall = true;
+    }
+
+    @Override
+    public void checkClassDecl(ClassDecl decl) {
+        for (ParamDecl p : decl.getParams()) {
+            setAnnotatedType(p.getType(), p, p.getAccess());
+        }
+
+        List<FieldDecl> nonNullFields = new ArrayList<>();
+        for (FieldDecl f : decl.getFields()) {
+            setAnnotatedType(f.getType(), f, f.getAccess());
+            if (f.nonNull() && !f.hasInitExp()) {
+                nonNullFields.add(f);
+            }
+        }
+
+        if (nonNullFields.isEmpty()) return;
+
+        if (!decl.hasInitBlock()) {
+            errors.add(new TypeError(
+                nonNullFields.get(0),
+                ErrorMessage.NULLABLE_TYPE_MISMATCH,
+                NullableType.NonNull.toString(),
+                NullableType.Nullable.toString()));
+            return;
+        }
+        // Get all fields that are nonNull at the end of the init block
+        BitVec<VarOrFieldDecl> out = decl.getInitBlock().exit().nonNull_in();
+        for (FieldDecl f : nonNullFields) {
+            if (!out.contains(f)) {
+                errors.add(new TypeError(
+                    f,
+                    ErrorMessage.NULLABLE_TYPE_MISMATCH,
+                    NullableType.NonNull.toString(),
+                    NullableType.Nullable.toString()));
+                // Only report one error
```

Comment:
If more than one non-nullable field is left null after the init block, do we report only one of them here?  And if so, can/should we report all of them?

Reply:
We definitely can. I was of the opinion that we shouldn't overwhelm the developer with similar errors. If they are nudged to consider that fields must be initialized, they may figure the concept out.

If you think we should report all errors, we definitely can.